### PR TITLE
Cleanup join distribution code

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 1]
```diff
@@ -91,22 +98,87 @@ public static boolean canReplicate(JoinNode joinNode, Context context)
         return buildSideSizeInBytes <= joinMaxBroadcastTableSize.toBytes();
     }
 
+    public static double getSourceTablesSizeInBytes(PlanNode node, Context context)
+    {
+        return getSourceTablesSizeInBytes(node, context.getLookup(), context.getStatsProvider(), context.getSymbolAllocator().getTypes());
+    }
+
+    @VisibleForTesting
+    static double getSourceTablesSizeInBytes(PlanNode node, Lookup lookup, StatsProvider statsProvider, TypeProvider typeProvider)
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java:107]: these are utilities, move them below where they are used

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 2]
```diff
@@ -91,22 +98,87 @@ public static boolean canReplicate(JoinNode joinNode, Context context)
...
+        boolean containsReplicatedJoin = possibleJoinNodes.stream()
+                .map(PlanNodeWithCost::getPlanNode)
+                .map(JoinNode.class::cast)
+                .anyMatch(node -> node.getDistributionType().equals(Optional.of(REPLICATED)));
+
+        JoinNode sizedBasedJoin = getSizeBasedJoin(joinNode, context);
+        if (!containsReplicatedJoin && sizedBasedJoin.getDistributionType().equals(Optional.of(REPLICATED))) {
+            // choose replicated size based join if stats calculator would skip replicated join altogether
+            // (because of inaccurate or missing intermediate plan node stats)
+            return sizedBasedJoin;
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java:140]: `containsReplicatedJoin` seems straighforward, based on `addJoinsWithDifferentDistributions` -- there is no cost decision yet.

i am concerned we exit too early here. 
Would it be sufficient to add this to `if (possibleJoinNodes.stream().anyMatch(result -> result.getCost().hasUnknownComponents()) || possibleJoinNodes.isEmpty())` branch below?

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 3]
```diff
@@ -91,22 +98,87 @@ public static boolean canReplicate(JoinNode joinNode, Context context)
...
+            return Double.NaN;
+        }
+
+        List<PlanNode> sourceNodes = PlanNodeSearcher.searchFrom(node, lookup)
+                .where(isInstanceOfAny(TableScanNode.class, ValuesNode.class))
+                .findAll();
+
+        return sourceNodes.stream()
+                .mapToDouble(sourceNode -> statsProvider.getStats(sourceNode).getOutputSizeInBytes(sourceNode.getOutputSymbols(), typeProvider))
+                .sum();
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java:122]: Add a test when one side eg UNION or UNION ALL of something

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java [hunk 4]
```diff
@@ -643,6 +655,185 @@ public void testReplicatesWhenNotRestricted()
                         values(ImmutableMap.of("B1", 0))));
     }
 
+    @Test
+    public void testReplicatesWhenSourceIsSmall()
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java:659]: Add "integration" test to `TestLogicalPlanner` as well.

File: presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java [hunk 5]
```diff
@@ -643,6 +655,185 @@ public void testReplicatesWhenNotRestricted()
...
+    {
+        VarcharType symbolType = createUnboundedVarcharType(); // variable width so that average row size is respected
+        int aRows = 10_000;
+        int bRows = 10;
+
+        // output size exceeds AUTOMATIC_RESTRICTED limit
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java:665]: is there such a constant?

Comment (sopel39) [reply presto-main/src/test/java/io/prestosql/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java:665]: no there isn't, but `AUTOMATIC_RESTRICTED` is used in other tests too. Let's make this cleanup a follow-up PR.

File: presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/DetermineJoinDistributionType.java:108]: The is one uitility above this one (`canReplicate`) and more utilities below, so i am not quite convinced. Up to you

File: presto-main/src/test/java/io/prestosql/sql/planner/TestLogicalPlanner.java [hunk 6]
```diff
@@ -1473,11 +1474,71 @@ public void testGroupingSetsWithDefaultValue()
...
+                                        anyTree(
+                                                tableScan("orders", ImmutableMap.of("CUSTKEY", "custkey"))),
+                                        anyTree(
+                                                values("T_A"))))));
+    }
+
+    @Test
+    public void testSizeBasedSemiJoin()
+    {
+        // both local.sf100000.nation and local.sf100000.orders don't provide stats, therefore no reordering happens
```
Comment (findepi) [inline presto-main/src/test/java/io/prestosql/sql/planner/TestLogicalPlanner.java:1506]: what will happen if we add stats for that schema?
(using `"sf42.5".orders` could buy you a bit more of safety)

**Review by findepi — APPROVED** [review]
one comment, optional

**Pull request merged**